### PR TITLE
Add frank-wolfe solver (v2)

### PR DESCRIPTION
**Pull request by antoinedemathelin**
Hi everyone,
Here is a new pull request for the frank-wolfe solver for KMM.
I made two qp_solve functions this time: `_qp_solve_scipy` and `_qp_solve_frank_wolfe`

### CODE REVIEW

File: skada/_reweight.py [hunk 1]
```diff
@@ -987,6 +987,8 @@ class KMMReweightAdapter(BaseAdapter):
         Number of maximum iteration before stopping the optimization.
     smooth_weights : bool, default=False
         If True, the weights are "smoothed" using the kernel function.
+    solver : string, default='qp_scipy'
```
Comment (rflamary) [inline skada/_reweight.py:990]:
```suggestion
    solver : string, default='frank-wolfe'
```

File: skada/tests/test_reweight.py [hunk 2]
```diff
@@ -78,6 +78,7 @@
         ),
         KMMReweight(),
         KMMReweight(eps=0.1),
+        KMMReweight(solver="frank-wolfe"),
```
Comment (rflamary) [inline skada/tests/test_reweight.py:81]: also add tes for scipy since fran-worlfe is now default

File: skada/utils.py [hunk 3]
```diff
@@ -558,9 +582,82 @@ def qp_solve(Q, c=None, A=None, b=None, Aeq=None, beq=None, lb=None, ub=None,
     max_iter : int, optional
         Maximum number of iterations to perform.
     verbose : boolean, optional
-        Print optimization information.
+        Print optimization informations.
```
Comment (rflamary) [inline skada/utils.py:585]: no s at the end of information

File: skada/utils.py [hunk 4]
```diff
@@ -628,6 +725,142 @@ def jac(x):
...
+        Maximum number of iterations to perform.
+
+    Returns
+    -------
+    x: (d,) ndarray
+        Optimal solution x
+    val: float
+        optimal value of the objective (None if optimization error)
+    """
+    if c is None:
```
Comment (rflamary) [inline skada/utils.py:784]: you should raise erreors here is A,b Aeq and beq do not respect the assumptions

File: skada/tests/test_utils.py [hunk 5]
```diff
@@ -506,6 +507,67 @@ def test_qp_solve():
...
+
+    res = qp_solve(Q, c, A=A2, b=b2,
+                   solver="frank-wolfe")
+    assert np.abs(res[0] - sol1).sum() < 0.01
+
+    res = qp_solve(Q, A=A2, b=b3, solver="frank-wolfe")
+    assert np.abs(res[0] - np.zeros(2)).sum() < 0.01
+
+    with pytest.raises(ValueError, match="`A` or `Aeq` must be given"):
+        qp_solve(Q, c, solver="frank-wolfe")
```
Comment (rflamary) [inline skada/tests/test_utils.py:546]: maybe test that scipy and rank-worlfe return th same vector?

**Review by rflamary — COMMENTED** [review]
Thanks @antoinedemathelin this is great and will definitely speedup several methods,

I have  a few comments below

File: skada/utils.py [hunk 6]
```diff
@@ -628,6 +725,142 @@ def jac(x):
...
+        grad = jac(x)
+        product = grad * inv_c
+        index = np.argmin(product)
+        vect = np.zeros(c.shape[0])
+        if product[index] >= 0:
+            vect[index] = inv_c[index] * clb
+        else:
+            vect[index] = inv_c[index] * cub
+        lr = 2. / (k + 1.)
+        x = (1 - lr) * x + lr * vect
```
Comment (agramfort) [inline skada/utils.py:860]: here you have no line search for lr. if f is quadratic you should be able to find the optimal step size at each iteration. But certainly for another PR.

File: skada/utils.py [hunk 7]
```diff
@@ -558,9 +582,82 @@ def qp_solve(Q, c=None, A=None, b=None, Aeq=None, beq=None, lb=None, ub=None,
...
+        \min_x \quad  \frac{1}{2}x^TQx + x^Tc
+
+
+        lb <= x <= ub
+
+        Ax <= b
+
+        A_{eq} x = b_{eq}
+
+    All constraint parameters are optional, they will be ignore is left at
```
Comment (agramfort) [inline skada/utils.py:630]:
```suggestion
    All constraint parameters are optional, they will be ignored if set as None.
```

File: skada/utils.py [hunk 8]
```diff
@@ -628,6 +725,142 @@ def jac(x):
...
+    -------
+    x: (d,) ndarray
+        Optimal solution x
+    """
+    inv_c = 1. / c
+
+    if x0 is None:
+        x0 = inv_c * ((cub - clb) / 2) / c.shape[0]
+
+    x = np.copy(x0)
```
Comment (agramfort) [inline skada/utils.py:848]: this copy is not necessary as you don't operate inplace on x below

**Pull request merged**